### PR TITLE
Use reuse tool 4.0

### LINKED ISSUE #127361: Upgrade to REUSE 4.x and `REUSE.toml`
[REUSE 4.0.0 was recently released][release], which addresses all the issues we had with REUSE through the addition of [`REUSE.toml`][toml]. We should upgrade the version of REUSE to it, and migrate from `.reuse/dep5` to `REUSE.toml` in the repository.

Notable things of `REUSE.toml` that would help our use:

* It is possible to instruct REUSE to ignore all annotations it would discover in source files and only use the annotations present in `REUSE.toml`. While in most cases we would *not* want to do that (as annotations in source files themselves is probably more accurate), we do want to override LLVM's licensing information with the boilerplate defined in `.reuse/dep5`/`REUSE.toml`. Setting `precedence = "override"` in `REUSE.toml` for LLVM only would address this.

* REUSE now supports having multiple `REUSE.toml`s in a repository, which means that a submodule can now define its own licensing annotations. Before that, only the top-level `.reuse/dep5` would be considered, and any other instance of that file in subdirectories or submodules would be ignored.

We should:

* [ ] Update the version of REUSE to 4.x.x
* [ ] Convert `.reuse/dep5` into `REUSE.toml`
* [ ] Revert https://github.com/rust-lang/rust/commit/c6eb03ba90f3d016a7fc75188182c08ce8281c40 and mark the LLVM annotation with `precedence = "override"`

Tracking issue: https://github.com/rust-lang/rust/issues/99414

[release]: https://github.com/fsfe/reuse-tool/releases/tag/v4.0.0
[toml]: https://reuse.software/spec-3.2/#reusetoml

### PR DESCRIPTION
**Pull request by jonathanpallant**
This change upgrades us to reuse-tool 4.0.3, which has a new TOML format configuration instead of the old `.reuse/dep5` Debian-style file.

* Updated requirements file to install reuse-4.0.3
* Ran `reuse convert-dep5` to switch to new file format
* Switched over to `override` so the `REUSE.toml` file takes precedence over whatever random Copyright strings `reuse` finds in the source tree.

Should fix https://github.com/rust-lang/rust/issues/127361

### CODE REVIEW

File: src/ci/docker/host-x86_64/mingw-check/reuse-requirements.txt [hunk 1]
```diff
@@ -1,9 +1,13 @@
 #
-# This file is autogenerated by pip-compile with Python 3.10
+# This file is autogenerated by pip-compile with Python 3.12
```
Comment (Kobzol) [inline src/ci/docker/host-x86_64/mingw-check/reuse-requirements.txt:2]: We use Python 3.10 in the Docker image of `mingw-check`. Could you please regenerate this lockfile with Python 3.10? Just to make sure that there are no surprises regarding Python package versions.

**Review by Kobzol — COMMENTED** [review]
This is great, thanks for working on it! I saw that you already discussed this with Pietro on the Ferrocene repo, and he is the author of the original issue, so he might want to bless this.

I cross-checked the TOML file with the original reuse file, everything looks fine. So apart from one nit (left a comment) LGTM.

r? @pietroalbini

File: src/ci/docker/host-x86_64/mingw-check/reuse-requirements.txt (hunk 1 above)
Comment (jonathanpallant) [reply src/ci/docker/host-x86_64/mingw-check/reuse-requirements.txt:2]: Done. The hashes were unchanged.

**Comment by pietroalbini** [comment]
Can you also revert https://github.com/rust-lang/rust/commit/c6eb03ba90f3d016a7fc75188182c08ce8281c40? With `precendence = "override"` set for LLVM, there should be no need for that manual tweaking anymore :D

**Comment by jonathanpallant** [comment]
That change has now been removed.

**Pull request merged**